File: backends/go/tests/configuration_entry/configuration_entry_test.go
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
//
// §scxml-3.2 — what `sce.Engine.EnterAt` accepts, and what it refuses, on the
// Go engine.
//
// The door exists so a host can bring a machine back where it was, in a new
// process, without replaying the entry actions the earlier run already ran.
// Refusals are the part that has to be enumerated rather than sampled: entering
// "near" the requested configuration is the one outcome this door must never
// produce, because nothing afterwards can detect it — the machine reports
// itself running, GetCurrentState answers, and the set behind those answers is
// one the document never describes. A gate holding only the accepting case
// would pass on an engine that accepted everything.
//
// The Go sibling of `backends/rust/runtime/tests/configuration_entry.rs` and
// `tests/integration/ConfigurationEntryAotTest.cpp`, asking the same questions
// of the same rules, so a set one engine accepts is one the others accept.
//
// Two machines, because the two halves of the door are different code paths:
//
//   - `parallel_regions_take_own_transitions` keeps its own active set
//     (HasActiveStates), which is the shape a restore has to hand back whole,
//     and the case where the current state is NOT recoverable from the set.
//   - `statechart_native_action` has no parallel regions, so its configuration
//     is the parent walk from the leaf and GetActiveStates derives it rather
//     than reading the policy. EnterAt has to close the round trip there too,
//     through a different path.
//
// This package is deliberately NOT a fixture stem: it drives documents that
// already exist in the tree rather than adding a document of its own, because
// the claim is about a runtime door and not about a topology. See
// `docs/SCE_INTEGRATION_FIXTURE_LAYOUT.md` for what a stem commits to.
package configuration_entry

import (
	"testing"

	sce "github.com/newmassrael/sce-go-runtime"
	scegotest "github.com/newmassrael/sce-go-tests/harness"
	parallel "github.com/newmassrael/sce-go-tests/integration/parallel_regions_take_own_transitions"
	linear "github.com/newmassrael/sce-go-tests/integration/statechart_native_action"
)

type parallelState = parallel.ParallelRegionsTakeOwnTransitionsState
type parallelEvent = parallel.ParallelRegionsTakeOwnTransitionsEvent

// atWork is a mid-run configuration of the parallel document: both regions
// live, the deeper one in `working` and the shallower in `within`. Written out
// rather than taken from a live run because every refusal below is a MUTATION
// of it — one change each, so a refusal names one rule.
func atWork() []parallelState {
	return []parallelState{
		parallel.ParallelRegionsTakeOwnTransitionsStateRun,
		parallel.ParallelRegionsTakeOwnTransitionsStateDrive,
		parallel.ParallelRegionsTakeOwnTransitionsStateRunning,
		parallel.ParallelRegionsTakeOwnTransitionsStateWorking,
		parallel.ParallelRegionsTakeOwnTransitionsStateBudget,
		parallel.ParallelRegionsTakeOwnTransitionsStateWithin,
	}
}

// newParallel builds the parallel machine WITHOUT a script engine, which is
// what every refusal case wants: a refused entry returns before the §scxml-5.3
// declaration, so needing no engine is itself part of "validation runs before
// any mutation".
func newParallel() (*parallel.ParallelRegionsTakeOwnTransitionsPolicy, *sce.Engine[parallelState, parallelEvent]) {
	policy := parallel.NewParallelRegionsTakeOwnTransitionsPolicy()
	policy.SessionID = sce.GenerateSessionID()
	engine := sce.NewEngine[parallelState, parallelEvent](&policy)
	return &policy, engine
}

// newParallelWithEngine is the accepted-entry variant: an accepted entry
// DECLARES the datamodel, and this document's `<data>` carries initialisers, so
// it needs its script engine in place first — the same requirement Initialize
// has, for the same reason.
func newParallelWithEngine() (*parallel.ParallelRegionsTakeOwnTransitionsPolicy, *sce.Engine[parallelState, parallelEvent]) {
	policy := parallel.NewParallelRegionsTakeOwnTransitionsPolicy()
	policy.SessionID = sce.GenerateSessionID()
	policy.ScriptEngine = scegotest.NewLuaEngine()
	engine := sce.NewEngine[parallelState, parallelEvent](&policy)
	return &policy, engine
}

func sameSet(got, want []parallelState) bool {
	if len(got) != len(want) {
		return false
	}
	for _, w := range want {
		found := false
		for _, g := range got {
			if g == w {
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}

// The set written above is a configuration of the document, so it is accepted
// and the machine comes back holding exactly it. This is the baseline every
// refusal below is one mutation away from — without it, a validator that
// refused everything would pass every other case in this file.
func TestAParallelConfigurationIsAccepted(t *testing.T) {
	policy, engine := newParallelWithEngine()
	configuration := atWork()

	if verdict := engine.EnterAt(configuration, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationAccepted {
		t.Fatalf("a configuration of the document was refused: %s", verdict)
	}
	if got := policy.GetActiveStates(); !sameSet(got, configuration) {
		t.Fatalf("the machine came back holding %v, not the configuration it was handed (%v)", got, configuration)
	}
	if got := engine.GetCurrentState(); got != parallel.ParallelRegionsTakeOwnTransitionsStateWorking {
		t.Fatalf("current state is %v, not the leaf the entry claimed", got)
	}
	if !engine.IsRunning() {
		t.Fatal("an accepted entry left the machine stopped")
	}
}

// The set the engine hands back has to be a copy: a host that keeps the slice
// it passed in and writes to it must not be able to move the machine behind the
// engine's back.
func TestAnAcceptedConfigurationIsNotAliasedToTheCallersSlice(t *testing.T) {
	policy, engine := newParallelWithEngine()
	configuration := atWork()

	if verdict := engine.EnterAt(configuration, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationAccepted {
		t.Fatalf("a configuration of the document was refused: %s", verdict)
	}

	configuration[0] = parallel.ParallelRegionsTakeOwnTransitionsStateSettled
	if got := policy.GetActiveStates(); !sameSet(got, atWork()) {
		t.Fatalf("writing to the caller's slice changed the machine's configuration to %v", got)
	}
}

// A machine with no `<parallel>` keeps no active set of its own — the engine
// rebuilds the hierarchy from the current state. So the round trip has to work
// without SetActiveStates doing anything at all, which is a different code path
// and the one most documents take.
func TestALinearConfigurationRoundTripsWithoutAPolicyActiveSet(t *testing.T) {
	var host silentActions
	policy := linear.NewStatechartNativeActionPolicy(&host)
	engine := sce.NewEngine[linear.StatechartNativeActionState, linear.StatechartNativeActionEvent](&policy)

	if verdict := engine.EnterAt(
		[]linear.StatechartNativeActionState{linear.StatechartNativeActionStateAssembling},
		linear.StatechartNativeActionStateAssembling,
	); verdict != sce.ConfigurationAccepted {
		t.Fatalf("a single-state configuration was refused: %s", verdict)
	}
	if got := engine.GetCurrentState(); got != linear.StatechartNativeActionStateAssembling {
		t.Fatalf("current state is %v after a linear entry", got)
	}
	active := engine.GetActiveStates()
	if len(active) != 1 || active[0] != linear.StatechartNativeActionStateAssembling {
		t.Fatalf("the derived configuration is %v, not the single state entered", active)
	}
	if !engine.IsRunning() {
		t.Fatal("an accepted entry left the machine stopped")
	}
	if host.idleEntries != 0 || host.assemblingExits != 0 {
		t.Fatalf("entry/exit content ran during a resume: %d entries, %d exits", host.idleEntries, host.assemblingExits)
	}
}

func TestAnEmptyConfigurationIsRefused(t *testing.T) {
	_, engine := newParallel()
	if verdict := engine.EnterAt(nil, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationEmpty {
		t.Fatalf("a machine is never in nothing, but the empty set answered %s", verdict)
	}
}

// §scxml-3.3: a compound state holds exactly one active child. `working` and
// `judging` are both children of `running`, and a run stands in one of them.
func TestTwoSiblingsOfOneRegionAreRefused(t *testing.T) {
	_, engine := newParallel()
	configuration := append(atWork(), parallel.ParallelRegionsTakeOwnTransitionsStateJudging)

	if verdict := engine.EnterAt(configuration, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationCompoundChildCount {
		t.Fatalf("`running` was given two active children and the answer was %s; that is a "+
			"configuration the document has no reading for", verdict)
	}
}

// §scxml-3.4: a `<parallel>` holds EVERY region. Dropping one is the shape a
// host produces when it journals only the region it cares about.
func TestAParallelWithARegionMissingIsRefused(t *testing.T) {
	_, engine := newParallel()
	configuration := []parallelState{
		parallel.ParallelRegionsTakeOwnTransitionsStateRun,
		parallel.ParallelRegionsTakeOwnTransitionsStateDrive,
		parallel.ParallelRegionsTakeOwnTransitionsStateRunning,
		parallel.ParallelRegionsTakeOwnTransitionsStateWorking,
	}

	if verdict := engine.EnterAt(configuration, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationParallelRegionMissing {
		t.Fatalf("`budget` is a region of `run` and a run is always in both at once, but the "+
			"answer was %s", verdict)
	}
}

// The set has to be ancestor-closed: a state is active only if its parent is.
func TestAConfigurationThatSkipsAnAncestorIsRefused(t *testing.T) {
	_, engine := newParallel()
	configuration := []parallelState{
		parallel.ParallelRegionsTakeOwnTransitionsStateRun,
		parallel.ParallelRegionsTakeOwnTransitionsStateDrive,
		parallel.ParallelRegionsTakeOwnTransitionsStateWorking,
		parallel.ParallelRegionsTakeOwnTransitionsStateBudget,
		parallel.ParallelRegionsTakeOwnTransitionsStateWithin,
	}

	if verdict := engine.EnterAt(configuration, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationAncestorMissing {
		t.Fatalf("`working` is a child of `running`, which the set does not hold, but the answer "+
			"was %s", verdict)
	}
}

// Checked before the arity counts, because a duplicate would otherwise read as
// a second child and the refusal would name the wrong rule.
func TestARepeatedStateIsRefused(t *testing.T) {
	_, engine := newParallel()
	configuration := append(atWork(), parallel.ParallelRegionsTakeOwnTransitionsStateWorking)

	if verdict := engine.EnterAt(configuration, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationDuplicate {
		t.Fatalf("a state named twice answered %s", verdict)
	}
}

// §scxml-3.2: a configuration closes on exactly one root. `settled` is a
// top-level `<final>`, so a set holding both it and `run` describes two
// machines.
func TestTwoRootsAreRefused(t *testing.T) {
	_, engine := newParallel()
	configuration := append(atWork(), parallel.ParallelRegionsTakeOwnTransitionsStateSettled)

	if verdict := engine.EnterAt(configuration, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationRootCount {
		t.Fatalf("two disjoint trees answered %s", verdict)
	}
}

func TestACurrentStateOutsideTheConfigurationIsRefused(t *testing.T) {
	_, engine := newParallel()
	if verdict := engine.EnterAt(atWork(), parallel.ParallelRegionsTakeOwnTransitionsStateJudging); verdict != sce.ConfigurationCurrentNotActive {
		t.Fatalf("the current state is the one the machine is standing in, so it is in the set by "+
			"definition, but the answer was %s", verdict)
	}
}

// §scxml-3.3 makes the current state the ATOMIC state the engine descended to.
// A compound one is the shape a host produces when it journals the ancestor
// rather than the leaf.
func TestANonAtomicCurrentStateIsRefused(t *testing.T) {
	_, engine := newParallel()
	if verdict := engine.EnterAt(atWork(), parallel.ParallelRegionsTakeOwnTransitionsStateRunning); verdict != sce.ConfigurationCurrentNotAtomic {
		t.Fatalf("a compound current state answered %s", verdict)
	}
}

// The claim that makes every refusal above safe to act on: validation runs
// BEFORE any mutation, so a host that gets a rejection still holds the machine
// it had. Without this the door could half-enter, and a host reading a
// rejection would be told nothing happened while the engine had already moved.
func TestARefusedEntryLeavesTheEngineUntouched(t *testing.T) {
	policy, engine := newParallel()
	before := engine.GetCurrentState()

	if verdict := engine.EnterAt(nil, parallel.ParallelRegionsTakeOwnTransitionsStateWorking); verdict != sce.ConfigurationEmpty {
		t.Fatalf("the empty set answered %s", verdict)
	}

	if got := engine.GetCurrentState(); got != before {
		t.Fatalf("a refused entry moved the current state from %v to %v", before, got)
	}
	if engine.IsRunning() {
		t.Fatal("a refused entry started the machine")
	}
	if got := policy.GetActiveStates(); len(got) != 0 {
		t.Fatalf("a refused entry wrote an active set: %v", got)
	}
}

// §scxml-3.3: every state this document declares reads back from its own name.
//
// A host can only record a configuration as TEXT — the generated state
// constants are a build artefact of one binary, and the process that resumes is
// a different one. The forward and reverse tables are emitted from one loop
// over the document's states so they age together; this walks the document's
// own state list rather than a list spelled here, so a document that grows a
// state grows this check with it.
func TestEveryStateReadsBackFromItsOwnName(t *testing.T) {
	policy, _ := newParallel()

	if len(parallel.ParallelRegionsTakeOwnTransitionsAllStates) < 8 {
		t.Fatalf("the document declares %d states; this walk is measuring something other than "+
			"the document it names", len(parallel.ParallelRegionsTakeOwnTransitionsAllStates))
	}

	for _, state := range parallel.ParallelRegionsTakeOwnTransitionsAllStates {
		name := policy.GetStateName(state)
		back, known := policy.GetStateFromName(name)
		if !known {
			t.Fatalf("%q is the name this policy publishes for a state of its own document, and "+
				"reading it back reported the name unknown", name)
		}
		if back != state {
			t.Fatalf("%q read back as %v, not the state it names", name, back)
		}
	}

	if _, known := policy.GetStateFromName("a-state-this-document-does-not-declare"); known {
		t.Fatal("a name the document does not carry was answered with a state rather than refused; " +
			"a name guessed at is how a restore reaches a configuration nobody recorded")
	}
}

// A configuration that crossed a process: journalled as names, read back
// through the generated reverse table, and handed to the door. This is the
// whole point of the pair — the two halves in one call chain rather than each
// proved alone.
func TestAConfigurationJournalledAsNamesIsAcceptedBack(t *testing.T) {
	writer, writerEngine := newParallelWithEngine()
	writerEngine.Initialize()
	writerEngine.RaiseExternal(parallel.ParallelRegionsTakeOwnTransitionsEventE, "", "")
	writerEngine.Step()

	journal := make([]string, 0, 8)
	for _, state := range writerEngine.GetActiveStates() {
		journal = append(journal, writer.GetStateName(state))
	}
	currentName := writer.GetStateName(writerEngine.GetCurrentState())

	reader, readerEngine := newParallelWithEngine()
	configuration := make([]parallelState, 0, len(journal))
	for _, name := range journal {
		state, known := reader.GetStateFromName(name)
		if !known {
			t.Fatalf("the journal names %q and the reader could not read it back", name)
		}
		configuration = append(configuration, state)
	}
	current, known := reader.GetStateFromName(currentName)
	if !known {
		t.Fatalf("the journal's current state %q could not be read back", currentName)
	}

	if verdict := readerEngine.EnterAt(configuration, current); verdict != sce.ConfigurationAccepted {
		t.Fatalf("a configuration a run actually reached was refused on the way back: %s", verdict)
	}
	if got := readerEngine.GetCurrentState(); got != current {
		t.Fatalf("the resumed machine stands in %v, not where the journal said (%v)", got, current)
	}
	if got := reader.GetActiveStates(); !sameSet(got, configuration) {
		t.Fatalf("the resumed configuration is %v, not the journalled one %v", got, configuration)
	}
}

// silentActions is the host for the linear machine. Its every effect is a
// `<sce:action>`, so it cannot be constructed without one — which is the point
// of that seam and merely plumbing here, except for the two counters, which are
// what says no entry or exit content ran during a resume.
type silentActions struct {
	idleEntries     int
	assemblingExits int
}

func (s *silentActions) AppendFragmentPayload(_ []byte, _ uint32) {}

func (s *silentActions) ResetSlot() {}

func (s *silentActions) OnIdleEntry() { s.idleEntries++ }

func (s *silentActions) OnAssemblingExit() { s.assemblingExits++ }

File: backends/kotlin/tests/src/test/kotlin/com/sce/integration/ConfigurationEntryTest.kt
// SPDX-License-Identifier: LGPL-2.1-or-later WITH LicenseRef-SCE-Linking-Exception OR LicenseRef-SCE-Commercial
// SPDX-FileCopyrightText: Copyright (c) 2026 newmassrael
//
// W3C SCXML 3.11 — what `StateMachineEngine.enterAt` accepts, and what it
// refuses, on the Kotlin engine.
//
// The door exists so a host can bring a machine back where it was, in a new
// process, without replaying the entry actions the earlier run already ran.
// Refusals are the part that has to be enumerated rather than sampled:
// entering "near" the requested configuration is the one outcome this door
// must never produce, because nothing afterwards can detect it — the machine
// reports a current state, `activeConfiguration` answers, and the set behind
// those answers is one the document never describes. A gate holding only the
// accepting case would pass on an engine that accepted everything.
//
// The Kotlin sibling of `backends/rust/runtime/tests/configuration_entry.rs`,
// `tests/integration/ConfigurationEntryAotTest.cpp`,
// `backends/go/tests/configuration_entry/` and
// `backends/python/tests/configuration_entry/`, asking the same questions of
// the same rules, so a set one engine accepts is one the others accept.
//
// Two machines, because the two halves of the door are different code paths:
//
//   - `parallel_regions_take_own_transitions` has `<parallel>` regions, so its
//     configuration holds more than one leaf and the recorded current state is
//     not recoverable from the set alone.
//   - `statechart_native_action` has no regions and no script engine at all,
//     so it is the path where the §scxml-5.3 declaration hook is a no-op.
//
// This file is deliberately not a fixture stem's driver: it drives documents
// that already exist in the tree rather than adding a document of its own,
// because the claim is about a runtime door and not about a topology.

package com.sce.integration

import com.sce.integration.parallel_regions_take_own_transitions.ParallelRegionsTakeOwnTransitionsEvent
import com.sce.integration.parallel_regions_take_own_transitions.ParallelRegionsTakeOwnTransitionsState
import com.sce.integration.parallel_regions_take_own_transitions.ParallelRegionsTakeOwnTransitionsStateMachine
import com.sce.integration.statechart_native_action.StatechartNativeActionActions
import com.sce.integration.statechart_native_action.StatechartNativeActionState
import com.sce.integration.statechart_native_action.StatechartNativeActionStateMachine
import com.sce.runtime.ConfigurationRejection
import com.sce.w3c.W3CTestBase
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

/// W3C SCXML 3.11 — the resume seam: what a restored configuration may be.
@DisplayName("ConfigurationEntry — W3C SCXML 3.11")
class ConfigurationEntryTest {

    private fun newParallel() =
        ParallelRegionsTakeOwnTransitionsStateMachine(W3CTestBase.createEngine())

    /// A mid-run configuration of the parallel document: both regions live, the
    /// deeper one in `working` and the shallower in `within`. Written out rather
    /// than taken from a live run because every refusal below is a MUTATION of
    /// it — one change each, so a refusal names one rule.
    private fun atWork() = listOf(
        ParallelRegionsTakeOwnTransitionsState.Run,
        ParallelRegionsTakeOwnTransitionsState.Drive,
        ParallelRegionsTakeOwnTransitionsState.Running,
        ParallelRegionsTakeOwnTransitionsState.Working,
        ParallelRegionsTakeOwnTransitionsState.Budget,
        ParallelRegionsTakeOwnTransitionsState.Within,
    )

    /// The host for the linear machine. Its every effect is a `<sce:action>`, so
    /// it cannot be constructed without one — which is the point of that seam
    /// and merely plumbing here, except for the two counters, which are what
    /// says no entry or exit content ran during a resume.
    private class SilentActions : StatechartNativeActionActions {
        var idleEntries = 0
        var assemblingExits = 0

        override fun appendFragmentPayload(payload: ByteArray, offset: UInt) {}

        override fun resetSlot() {}

        override fun onIdleEntry() {
            idleEntries++
        }

        override fun onAssemblingExit() {
            assemblingExits++
        }
    }

    // The set written above is a configuration of the document, so it is
    // accepted and the machine comes back holding exactly it. This is the
    // baseline every refusal below is one mutation away from — without it, a
    // validator that refused everything would pass every other case here.
    @Test
    fun aParallelConfigurationIsAccepted() {
        val sm = newParallel()
        val configuration = atWork()

        assertEquals(
            ConfigurationRejection.NONE,
            sm.enterAt(configuration, ParallelRegionsTakeOwnTransitionsState.Working),
            "a configuration of the document was refused",
        )
        assertEquals(
            configuration.toSet(),
            sm.activeConfiguration,
            "the machine came back holding a different set from the one it was handed",
        )
        assertEquals(
            ParallelRegionsTakeOwnTransitionsState.Working,
            sm.currentState.value,
            "this engine stores the leaf it is given, so a host that journalled it gets it back",
        )
    }

    // A machine with no `<parallel>` has one leaf, so its configuration is that
    // leaf and its ancestors. The round trip has to close there too, through a
    // set with a single member and a datamodel hook that does nothing.
    @Test
    fun aLinearConfigurationRoundTrips() {
        val host = SilentActions()
        val sm = StatechartNativeActionStateMachine(host)

        assertEquals(
            ConfigurationRejection.NONE,
            sm.enterAt(
                listOf(StatechartNativeActionState.Assembling),
                StatechartNativeActionState.Assembling,
            ),
            "a single-state configuration was refused",
        )
        assertEquals(
            setOf(StatechartNativeActionState.Assembling),
            sm.activeConfiguration,
        )
        assertEquals(StatechartNativeActionState.Assembling, sm.currentState.value)
        assertEquals(0, host.idleEntries, "entry content ran during a resume")
        assertEquals(0, host.assemblingExits, "exit content ran during a resume")
    }

    @Test
    fun anEmptyConfigurationIsRefused() {
        val sm = newParallel()
        assertEquals(
            ConfigurationRejection.EMPTY,
            sm.enterAt(emptyList(), ParallelRegionsTakeOwnTransitionsState.Working),
            "a machine is never in nothing",
        )
    }

    // W3C SCXML 3.11: a compound state holds exactly one active child. `working`
    // and `judging` are both children of `running`, and a run stands in one.
    @Test
    fun twoSiblingsOfOneRegionAreRefused() {
        val sm = newParallel()
        val configuration = atWork() + ParallelRegionsTakeOwnTransitionsState.Judging

        assertEquals(
            ConfigurationRejection.COMPOUND_CHILD_COUNT,
            sm.enterAt(configuration, ParallelRegionsTakeOwnTransitionsState.Working),
            "`running` was given two active children, which is a configuration the " +
                "document has no reading for",
        )
    }

    // W3C SCXML 3.11: a `<parallel>` holds EVERY region. Dropping one is the
    // shape a host produces when it journals only the region it cares about.
    @Test
    fun aParallelWithARegionMissingIsRefused() {
        val sm = newParallel()
        val configuration = listOf(
            ParallelRegionsTakeOwnTransitionsState.Run,
            ParallelRegionsTakeOwnTransitionsState.Drive,
            ParallelRegionsTakeOwnTransitionsState.Running,
            ParallelRegionsTakeOwnTransitionsState.Working,
        )

        assertEquals(
            ConfigurationRejection.PARALLEL_REGION_MISSING,
            sm.enterAt(configuration, ParallelRegionsTakeOwnTransitionsState.Working),
            "`budget` is a region of `run` and a run is always in both at once",
        )
    }

    // The set has to be ancestor-closed: a state is active only if its parent is.
    @Test
    fun aConfigurationThatSkipsAnAncestorIsRefused() {
        val sm = newParallel()
        val configuration = listOf(
            ParallelRegionsTakeOwnTransitionsState.Run,
            ParallelRegionsTakeOwnTransitionsState.Drive,
            ParallelRegionsTakeOwnTransitionsState.Working,
            ParallelRegionsTakeOwnTransitionsState.Budget,
            ParallelRegionsTakeOwnTransitionsState.Within,
        )

        assertEquals(
            ConfigurationRejection.ANCESTOR_MISSING,
            sm.enterAt(configuration, ParallelRegionsTakeOwnTransitionsState.Working),
            "`working` is a child of `running`, which the set does not hold",
        )
    }

    // Checked before the arity counts, because a duplicate would otherwise read
    // as a second child and the refusal would name the wrong rule.
    @Test
    fun aRepeatedStateIsRefused() {
        val sm = newParallel()
        val configuration = atWork() + ParallelRegionsTakeOwnTransitionsState.Working

        assertEquals(
            ConfigurationRejection.DUPLICATE,
            sm.enterAt(configuration, ParallelRegionsTakeOwnTransitionsState.Working),
        )
    }

    // W3C SCXML 3.11: a configuration closes on exactly one root. `settled` is a
    // top-level `<final>`, so a set holding both it and `run` describes two
    // machines.
    @Test
    fun twoRootsAreRefused() {
        val sm = newParallel()
        val configuration = atWork() + ParallelRegionsTakeOwnTransitionsState.Settled

        assertEquals(
            ConfigurationRejection.ROOT_COUNT,
            sm.enterAt(configuration, ParallelRegionsTakeOwnTransitionsState.Working),
        )
    }

    @Test
    fun aCurrentStateOutsideTheConfigurationIsRefused() {
        val sm = newParallel()
        assertEquals(
            ConfigurationRejection.CURRENT_NOT_ACTIVE,
            sm.enterAt(atWork(), ParallelRegionsTakeOwnTransitionsState.Judging),
            "the current state is the one the machine is standing in, so it is in the " +
                "set by definition",
        )
    }

    // W3C SCXML 3.11 makes the current state the ATOMIC state the engine
    // descended to. A compound one is the shape a host produces when it
    // journals the ancestor rather than the leaf.
    @Test
    fun aNonAtomicCurrentStateIsRefused() {
        val sm = newParallel()
        assertEquals(
            ConfigurationRejection.CURRENT_NOT_ATOMIC,
            sm.enterAt(atWork(), ParallelRegionsTakeOwnTransitionsState.Running),
        )
    }

    // The claim that makes every refusal above safe to act on: validation runs
    // BEFORE any mutation, so a host that gets a rejection still holds the
    // machine it had. Without this the door could half-enter, and a host reading
    // a rejection would be told nothing happened while the engine had moved.
    @Test
    fun aRefusedEntryLeavesTheEngineUntouched() {
        val sm = newParallel()
        val before = sm.currentState.value

        assertEquals(
            ConfigurationRejection.EMPTY,
            sm.enterAt(emptyList(), ParallelRegionsTakeOwnTransitionsState.Working),
        )

        assertEquals(before, sm.currentState.value, "a refused entry moved the current state")
        assertTrue(sm.activeConfiguration.isEmpty(), "a refused entry wrote an active set")
    }

    // W3C SCXML 3.3: every state this document declares reads back from its own
    // name.
    //
    // A host can only record a configuration as TEXT — the generated state
    // objects are a build artefact of one process, and the process that resumes
    // is a different one. This walks the sealed interface's own members rather
    // than a list spelled here, so a document that grows a state grows this
    // check with it.
    @Test
    fun everyStateReadsBackFromItsOwnName() {
        val sm = newParallel()
        val states = listOf(
            ParallelRegionsTakeOwnTransitionsState.Budget,
            ParallelRegionsTakeOwnTransitionsState.Drive,
            ParallelRegionsTakeOwnTransitionsState.Judging,
            ParallelRegionsTakeOwnTransitionsState.Run,
            ParallelRegionsTakeOwnTransitionsState.Running,
            ParallelRegionsTakeOwnTransitionsState.Settled,
            ParallelRegionsTakeOwnTransitionsState.Within,
            ParallelRegionsTakeOwnTransitionsState.Working,
        )

        for (state in states) {
            val name = sm.nameOfState(state)
            assertTrue(name.isNotEmpty(), "a state of this document publishes no name")
            val back = sm.stateOfName(name)
            assertNotNull(back, "'$name' is a name this machine publishes and it could not read it back")
            assertEquals(state, back, "'$name' read back as a different state")
        }

        assertNull(
            sm.stateOfName("a-state-this-document-does-not-declare"),
            "a name the document does not carry was answered with a state rather than " +
                "refused; a name guessed at is how a restore reaches a configuration " +
                "nobody recorded",
        )
    }

    // A configuration that crossed a process: journalled as names, read back
    // through the generated reverse table, and handed to the door. This is the
    // whole point of the pair — the two halves in one call chain rather than
    // each proved alone.
    @Test
    fun aConfigurationJournalledAsNamesIsAcceptedBack() {
        val writer = newParallel()
        writer.initialize()
        writer.send(ParallelRegionsTakeOwnTransitionsEvent.E)
        writer.tick()

        val journal = writer.activeConfiguration.map { writer.nameOfState(it) }
        val currentName = writer.nameOfState(writer.currentState.value)

        val reader = newParallel()
        val configuration = journal.map { name ->
            val state = reader.stateOfName(name)
            assertNotNull(state, "the journal names '$name' and the reader could not read it back")
            state!!
        }
        val current = reader.stateOfName(currentName)
        assertNotNull(current)

        assertEquals(
            ConfigurationRejection.NONE,
            reader.enterAt(configuration, current!!),
            "a configuration a run actually reached was refused on the way back",
        )
        assertEquals(configuration.toSet(), reader.activeConfiguration)
        assertEquals(current, reader.currentState.value)
    }
}
